test/refunder.ts

```diff
@@ -0,0 +1,299 @@
+
+			// address, bytes4
+			const resBefore = await refunder.refundables(randomAddress, randomFuncIdAsBytes); 
+			expect(resBefore).to.be.eq(false);
+
+			let res = await refunder.whitelistRefundable(randomAddress, randomFuncIdAsBytes, true);
+			await res.wait();
+
+			const resAfter = await refunder.refundables(randomAddress, randomFuncIdAsBytes); 
+			expect(resAfter).to.be.eq(true);
+
+			res = await refunder.whitelistRefundable(randomAddress, randomFuncIdAsBytes, false);
+			await res.wait();
+
+			const resAfterEdit = await refunder.refundables(randomAddress, randomFuncIdAsBytes); 
+			expect(resAfterEdit).to.be.eq(false);
+		});
+
+		it("Not owner should NOT be able to add refundable", async function() {
+		
+			const randomAddress = greeter.address;
+			const randomFuncId = ethers.utils.id('setGreeting(string)');
+			const randomFuncIdAsBytes = ethers.utils.arrayify(randomFuncId.substr(0, 10))
+
+			await expect(refunder.connect(addr1).whitelistRefundable(randomAddress, randomFuncIdAsBytes, true)).to.be.revertedWith(NOT_AN_OWNER);
+		});
+
+		it("Not owner should NOT be able to edit refundable", async function() {
+			const randomAddress = greeter.address;
+			const randomFuncIdAsBytes = generateFuncIdAsBytes('setGreeting(string)');
+
+			let res = await refunder.whitelistRefundable(randomAddress, randomFuncIdAsBytes, true);
+			await res.wait();
+
+			await expect(refunder.connect(addr1).whitelistRefundable(randomAddress, randomFuncIdAsBytes, false)).to.be.revertedWith(NOT_AN_OWNER);
+		});
+	});
+
+	describe('Relay and refund', () => {
+
+		beforeEach(async () => {
+			
+			let res = await owner.sendTransaction({
+				value: ethToWei("1"),
+				to: refunder.address
+			});
+
+			await res.wait();
+
+			res = await refunder.setMaxGasPrice('150000000000'); // 150 gwei
+			await res.wait();
+		});
+
+		it('Successfully refund whitelisted contract function', async () => {
+
+			const funcIdAsBytes = generateFuncIdAsBytes('setGreeting(string)');
+			
+			let res = await refunder.whitelistRefundable(greeter.address, funcIdAsBytes, true);
+			await res.wait();
+
+			const balanceBefore = await ethers.provider.getBalance(addr1.address);
+			
+			const text = 'Hello, Tester!';
+			const hexString = strToHex(text);
+			const args = ethers.utils.arrayify(hexString);
+			
+			res = await refunder.connect(addr1).relayAndRefund(greeter.address, funcIdAsBytes, args, {
+				gasLimit: 200000
+			});
+
+			let txReceipt = await res.wait();
+			
+			expect(txReceipt.events, 'No events are emitted').to.be.ok;
+			expect(txReceipt.events[0].event, 'Invalid event name').to.be.eq('RelayAndRefund');
+
+			const balanceAfter = await ethers.provider.getBalance(addr1.address);
+
+			const difference = balanceBefore.sub(balanceAfter);
+			expect(difference.lt(BigNumber.from('200000000000')), `Account was not refunded or refunded too much. ${difference.toString()}`).to.be.ok; // 200 gwei
```

Comment:
Here we are hardcoding `200000000000`. I think that it will be more precise if we do the following:

1. Calculate the `ETH cost of the TX` (`gasPrice` * `gasUsed`)
2. Check that `balanceBefore <  balanceAfter` (checks that we are not refunding more than needed)
3. Check that `balanceBefore - balanceAfter < 0.05*{Actual Cost of ETH TX} (this would check that the user did not pay more than 5% of the TX. We refund him at least 95% of the costs